### PR TITLE
cache: keep & serve fetched chunks in memory & configurable disk write concurrency

### PR DESCRIPTION
**Pull request by danielnorberg**
* Cache pending disk cache write chunks in ram to avoid re-fetching and read-amplification over congested origin links (e.g VPC tunnels). 
  * Configurable by `--diskcache-pending-write-cache-limit-bytes` flag and `BLOBNET_DISKCACHE_WRITE_CACHE_LIMIT_BYTES` env var.
  * Serve `get` requests from the pending disk cache write cache to avoid re-fetching for subsequent duplicate requests while chunk is pending write to disk.
  * This re-uses the in-memory `PageCache` data structure to hold a limited volume of chunks that are pending writing to disk cache, although in a separate instance and size limit from the pre-existing `PageCache`. The pending writes `PageCache` contents are removed after writing to disk, where it can be better cached by the OS-provided page cache, while allowing the server process heap to expand to hold and serve data that has yet not been written to disk.
* Configurable disk cache population write concurrency by `--diskcache-write-concurrency` flag and `BLOBNET_DISKCACHE_WRITE_CONCURRENCY` env var. 
  * Server default raised from `1` to `4`
* Add a `CacheConfig` with builder for configuration of the `Cached` provider as the number of parameters is growing and backwards compatibility hard to maintain in the constructor. 
* Have a fixed set of dedicated long-lived threads perform disk writes instead of spawning tasks per disk write as that seems to suffer from starvation when the blobnet server is taking traffic.
  * As a bonus this also fixes the problem where the blobnet server would refuse to shut down on SIGTERM.

This change has been stress tested by running the server on a `c6id.8xlarge` ec2 instance (cache on local SSD, up to 12.5 Gbit/s network bandwidth) with `--diskcache-pending-write-cache-limit-bytes 4294967296 --diskcache-write-concurrency 4`. A separate `m6i.2xlarge` instance sends requests at varying concurrency rates (100 - 2000). Whereas the current implementation fails to flush to disk, resulting in ever-growing `pending_disk_write_{pages,bytes}` metrics, with this PR these metrics stay just above 0, meaning that the disk cache population is able to keep up with the sustained traffic.

### Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: blobnet/src/provider.rs [hunk 1]
```diff
@@ -643,79 +768,88 @@ impl<P: Provider + 'static> CachedState<P> {
...
-
-        let state = scopeguard::guard(self.clone(), move |state| {
-            state.diskcache_pending_write_pages.fetch_sub(1, Relaxed);
-            state
-                .diskcache_pending_write_bytes
-                .fetch_sub(bytes_len, Relaxed);
+    async fn populate_disk_cache(self: &Arc<Self>, hash: String, n: u64, bytes: Bytes) {
+        let bytes_len = bytes.len() as u64;
+
+        // TODO (dano): have disk cache populator to re-fetch chunks that are evicted
```
Comment (gimaker) [inline blobnet/src/provider.rs:782]: Why evict stuff in the first place?

Wouldn't it be better to refetch the chunk if in-memory buffer for pending writes is full? I.e. keep the current behaviour if the cache is full.

Comment (danielnorberg) [reply blobnet/src/provider.rs:782]: Unless we evict we might eventually run out of memory if writing to disk is slow and unable to keep up with the throughput. Also, in workers we have observed that blowing up the process heap with pending chunks significantly negatively impacts cold start latency.

File: blobnet/src/provider.rs [hunk 2]
```diff
@@ -471,26 +506,108 @@ pub struct CacheStats {
...
+        // Note: We spawn dedicated long-lived threads to perform disk cache writes.
+        // Previously we spawned blocking tasks per chunk to flush it to disk,
+        // but this seemed to suffer from starvation when blobnet was also
+        // serving requests. The pending disk cache writes would queue up but
+        // forward progress writing them to disk was very slow. The root cause is still
+        // unclear but dedicated threads seem to avoid this problem.
+        for _ in 0..config.diskcache_write_concurrency {
+            let state = Arc::clone(&state);
+            let rx = diskcache_write_queue_rx.clone();
+            std::thread::spawn(move || {
```
Comment (gimaker) [inline blobnet/src/provider.rs:574]: Are these daemon threads? Wondering if `_disk_cache_populator` can be killed mid-blob-write on shutdown :thinking: 

I.e. program exits before it processes the shutdown signal

Comment (danielnorberg) [reply blobnet/src/provider.rs:572]: https://doc.rust-lang.org/std/thread/#the-threading-model

> When the main thread of a Rust program terminates, the entire program shuts down, even if other threads are still running.

Comment (danielnorberg) [reply blobnet/src/provider.rs:572]: Yeah, it is now actually atomic since a few weeks back so shutting down mid-write should be fine 😅

**Pull request merged**